feat: cache auth tokens in GenServer per connection

elliottneilclark opened the pull request:
<!-- Describe your pull request here. !-->
When connecting to the kube api there are often authentication data that's needed for each request. The lifecycle of that authentication token data is important to stability and performance. This pull request will add on running processes that cache and refresh tokens and certificates.

Tokens can be generated via a process. Shelling out will return a [ExecCredentialStatus](https://kubernetes.io/docs/reference/config-api/client-authentication.v1/#client-authentication-k8s-io-v1-ExecCredentialStatus). The ExecWorker added in this PR will cache that value (meaning that subsequent requests will not need to fork out many times). Additionally this PR will implement a refresh on a timer. That has the added benefit of making distributed systems much more reliable as we are able to retry a few times befor expiary, and we can add jitter.

Certificates in cloud envioriments can be ephemeral. They are not created and then kept forever. Often they have lifespans in the minutes. This PR will add on a CertificateWorker that caches the value from disk, and refresh it before the timeout.

---

<!-- In order for this pull request to be merged it has to fulfill the following requirements: -->

#### Requirements for all pull requests

- [x] Entry in CHANGELOG.md was created

#### Additional requirements for new features

- [x] New unit tests were created
- [x] New integration tests were created
- [x] PR description contains a link to the according kubernetes documentation

elliottneilclark committed:
feat: start a dynamic supervisor for auth providers

Summary:

- Start `K8s.Conn.Auth.ProviderSupervisor`
- Start `K8s.Conn.Auth.ExecWorker` for kube context with auth exec
  configured. This respects expiration and will not give out known
  expired tokens.
- Start `K8s.Conn.Auth.ServiceAccountWorker` for connections started
  with `K8s.Conn.from_service_account/2`. This periodically refreshes
  the token.
- Add refresh timers with jitter on for all of those
- Start `K8s.Conn.Auth.CertificateWorker` for kube context where given a
  .pem file. Some cloud providers are giving very short duration certs.

Test Plan:

- Added test for exec parsing
- Added test for exec worker
- Added test for service account
- Added integration-ish test for service account worker refresh
- Changed test for certificate to ensure genserver
- Added test for certificate worker

`mix test`

mruoss commented on lib/k8s/conn/auth/exec_worker.ex:
```diff
@@ -0,0 +1,128 @@
+
+    @type t :: %__MODULE__{
+            command: String.t(),
+            env: %{name: String.t(), value: String.t()},
```
I know this is copy pasted, but looking at the code, I think this spec should be for `%{name: "value"}`, right?

```suggestion
            env: %{name: String.t()},
```

mruoss commented on lib/k8s/conn/auth/exec_worker.ex:
```diff
@@ -0,0 +1,215 @@
+      timer: nil
+    ]
+
+    @type t :: %__MODULE__{
```
For further development it might help to have a hint to where this structure is coming from, wdyt?

```suggestion
    @typedoc """ 
    Simplified version of [ExecConfig](https://kubernetes.io/docs/reference/config-api/kubeconfig.v1/#ExecConfig)
    """
    @type t :: %__MODULE__{
```

elliottneilclark replied:
Perfect, thank you. I'll add some more typedocs for state as well.

JTarasovic commented on lib/k8s/conn/auth/exec_worker.ex:
```diff
@@ -0,0 +1,215 @@
+  defp format_env(env), do: Map.new(env, &{&1["name"], &1["value"]})
+
+  @spec valid_token?(State.t()) :: boolean()
+  defp valid_token?(%State{token: token, expiration_timestamp: exp} = _state) do
```
This may be clearer to express as guards?

e.g.
```elixir
defp valid_token?(%State{token: token} = _state) when is_nil(token), do: false
defp valid_token?(%State{token: token} = _state) when byte_size(token) == 0, do: false
defp valid_token?(%State{token: token, expiration_timestamp: exp} = _state) when is_nil(exp), do: true
defp valid_token?(%State{token: token, expiration_timestamp: exp} = _state), do: DateTime.after?(exp, DateTime.utc_now())
```

elliottneilclark replied:
That is way better.

JTarasovic commented on lib/k8s/conn/auth/exec_worker.ex:
```diff
@@ -0,0 +1,215 @@
+       }) do
+    case DateTime.from_iso8601(expire) do
+      {:ok, expiration_timestamp, _} ->
+        {:ok, token: token, expiration_timestamp: expiration_timestamp}
```
Does it make sense to return this as a map or tuple? Then we wouldn't have to `Keyword.get` elsewhere.

JTarasovic reviewed (commented):
Quick initial review.

mruoss commented on lib/k8s/conn.ex:
```diff
@@ -219,7 +220,7 @@ defmodule K8s.Conn do
       conn = %Conn{
         url: kubernetes_service_url(),
         ca_cert: ca_cert,
-        auth: %K8s.Conn.Auth.Token{token: token},
```
Is `K8s.Conn.Auth.Token` still needed or can we delete it?

elliottneilclark replied:
Token is still used by DigitalOcean, and integration testing. We don't have a do account, so I didn't want to break those.

mruoss reviewed (commented):
already looks very nice!

JTarasovic commented on test/k8s/conn/auth/exec_worker_test.exs:
```diff
@@ -0,0 +1,119 @@
+  alias K8s.Conn.Auth.ExecWorker
+
+  describe "parse_opts/1" do
+    test "parse a comman with nothing else" do
```
```suggestion
    test "parse a command with nothing else" do
```

JTarasovic commented on test/k8s/conn/auth/exec_worker_test.exs:
```diff
@@ -0,0 +1,119 @@
+    test "parse a command with arguments" do
+      config = %{
+        "exec" => %{
+          "apiVersion" => "client.authentication.k8s.io/v1alpha1",
```
It probably doesn't matter all that much but v1alpha1 has been removed for quite a while. I don't know that there's a supported version that still has it? Perhaps `v1` would make more sense in that case?

elliottneilclark committed:
fix: allow auth genserver to crash and restart

Summary:
This adds on a registry and a standard name for many of the auth
providers. It allows for shared workers when the paths are the same and
for genserver to crash, get re-started and not have errors due to stale
pids.

Test Plan:
Tests were updated

mruoss commented:
Hey @elliottneilclark, is this PR ready for review?

mruoss commented on lib/k8s/conn/auth/certificate.ex:
```diff
@@ -35,7 +57,19 @@ defmodule K8s.Conn.Auth.Certificate do
+        {:ok,
+         %K8s.Conn.RequestOptions{
+           headers: [],
+           ssl_options: [cert: cert, key: key]
```
Instead of adding a CertificateWorker GenServer, could we just use [`certfile` and `keyfile`](https://www.erlang.org/doc/man/ssl#type-common_option)? Or do you see this as performance optimisation by reducing File IO? But then again we're keeping HTTP2 conns open if possible...

mruoss commented on lib/k8s/conn/auth/exec.ex:
```diff
@@ -31,42 +31,45 @@ defmodule K8s.Conn.Auth.Exec do
+    # We don't want to be in the habit of parsing command names for unique identifiers
+    # so every auth worker gets it's own unique name. Then we use that name to
+    # register the worker with the registry.
+    id = :crypto.strong_rand_bytes(16) |> Base.url_encode64(padding: false)
```
How about creating an md5 hash of the args and use it as key in the registry? 

Otherwise, if you spin up a new ExecWorker every time, I don't get what you need a Registry for. Can't you start the genserver under the dynamic supervisor and register the pid right here in the struct? I.e. return `{:ok, %__MODULE__{target: pid}}`

elliottneilclark replied:
The registry and naming allow for GenServers to crash and still have a working `Conn`.

Imagine this sequence.

- User creates a new K8s.Conn
- The Exec worker starts with a pid of 12 since it's the 12th process.
- `K8s.Conn` needs a token so it sends a request to pid 12
- Something strange happens with calling the exec binary (not sure what, but something)
- The exec worker gets restarted with a pid of 13 by the supervisor
- The `K8s.Conn` needs a token so it sends pid 12 a request.
- The `K8s.Conn` is broken and won't work again.

With the name and the registry we get a way to look that up. So the sequence with the current strucutre would be:


- User creates a new K8s.Conn
- The Exec worker starts with a pid of 12 since it's the 12th process.
- The registry gets a mapping of name to pid.
- `K8s.Conn` needs a token
- We ask registry for a pid to send a request
- We send the request to pid 12
- Something strange happens with calling the exec binary (not sure what, but something)
- The exec worker gets restarted with a pid of 13 by the supervisor
- The registry gets the new mapping of name to pid 13
- The `K8s.Conn` needs a token so it  checks the registry and gets pid 13
- The `K8s.Conn` works after the process restarts.

mruoss replied:
Yes you are right of course. Thanks for the detailed explanations.

mruoss commented on lib/k8s/conn/auth/service_account.ex:
```diff
@@ -0,0 +1,64 @@
+  @type t :: %__MODULE__{target: GenServer.server()}
+
+  @impl true
+  @spec create(map() | any, String.t() | any) :: {:ok, t} | {:error, Error.t() | any()} | :skip
```
map() | any is just any, no?

elliottneilclark committed:
refactor: remove cerficate worker

elliottneilclark committed:
fix: K8s.Conn.Auth.ServiceAccount type spec

elliottneilclark commented:
@mruoss All the review comments are addressed. This should be good to go.

mruoss commented:
very nice, thanks @elliottneilclark

elliottneilclark commented:
@mruoss Any chance of a released version? I would love to not point at git :-)

mruoss commented:
ouff absolutely! Sorry about that!

elliottneilclark commented:
<img src="https://media2.giphy.com/media/KJ1f5iTl4Oo7u/giphy-downsized-medium.gif"/>
